File: GamePad-Mobile/mobile/lib/core/protocol/packet_header.dart
import 'enums.dart';

/// Fixed 8-byte header for every GamePad protocol packet.
class PacketHeader {
  final int version;
  final MessageType type;
  final int sequenceNumber;
  final int timestampMs;

  const PacketHeader({
    required this.version,
    required this.type,
    required this.sequenceNumber,
    required this.timestampMs,
  });

  /// Serialises this header into [dest] (must be at least 8 bytes).
  void writeTo(List<int> dest) {
    dest[0] = version;
    dest[1] = type.value;
    dest[2] = (sequenceNumber >> 8) & 0xFF;
    dest[3] = sequenceNumber & 0xFF;
    dest[4] = (timestampMs >> 24) & 0xFF;
    dest[5] = (timestampMs >> 16) & 0xFF;
    dest[6] = (timestampMs >> 8) & 0xFF;
    dest[7] = timestampMs & 0xFF;
  }

  /// Reads a header from [src] (must be at least 8 bytes).
  static PacketHeader readFrom(List<int> src) {
    return PacketHeader(
      version: src[0],
      type: MessageType.fromValue(src[1]),
      sequenceNumber: (src[2] << 8) | src[3],
      timestampMs: (src[4] << 24) | (src[5] << 16) | (src[6] << 8) | src[7],
    );
  }

  static const int byteSize = 8;

  @override
  String toString() =>
      'v$version type=$type seq=$sequenceNumber ts=${timestampMs}ms';
}

File: GamePad-Mobile/pc-wpf/Protocol/PacketHeader.cs
namespace GamePadServer.Wpf.Protocol;

/// <summary>
/// Fixed 8-byte header for every GamePad protocol packet.
/// </summary>
public readonly struct PacketHeader
{
    /// <summary>Protocol version (currently 0x01).</summary>
    public byte Version { get; }

    /// <summary>Message type identifying the payload.</summary>
    public MessageType Type { get; }

    /// <summary>Monotonic sequence number (per sender).</summary>
    public ushort SequenceNumber { get; }

    /// <summary>Timestamp in milliseconds since session start.</summary>
    public uint TimestampMs { get; }

    public PacketHeader(byte version, MessageType type, ushort sequenceNumber, uint timestampMs)
    {
        Version = version;
        Type = type;
        SequenceNumber = sequenceNumber;
        TimestampMs = timestampMs;
    }

    /// <summary>
    /// Serialises the header to a byte span (big-endian).
    /// </summary>
    public void WriteTo(Span<byte> dest)
    {
        dest[0] = Version;
        dest[1] = (byte)Type;
        dest[2] = (byte)(SequenceNumber >> 8);
        dest[3] = (byte)(SequenceNumber);
        dest[4] = (byte)(TimestampMs >> 24);
        dest[5] = (byte)(TimestampMs >> 16);
        dest[6] = (byte)(TimestampMs >> 8);
        dest[7] = (byte)(TimestampMs);
    }

    /// <summary>
    /// Reads a header from a byte span (big-endian).
    /// </summary>
    public static PacketHeader ReadFrom(ReadOnlySpan<byte> src)
    {
        return new PacketHeader(
            version: src[0],
            type: (MessageType)src[1],
            sequenceNumber: (ushort)((src[2] << 8) | src[3]),
            timestampMs: (uint)((src[4] << 24) | (src[5] << 16) | (src[6] << 8) | src[7])
        );
    }

    /// <summary>
    /// Size of the serialised header in bytes.
    /// </summary>
    public const int ByteSize = 8;

    public override string ToString() =>
        $"v{Version} type={Type} seq={SequenceNumber} ts={TimestampMs}ms";
}
